src/meanshift.jl

```diff
@@ -0,0 +1,96 @@
+"""
+```
+segments                = meanshift(img, spatial_radius, range_radius, [iter], [eps])
+``` 
+Segments the image using meanshift clustering.
+    
+Parameters:  
+-    img                            = input grayscale image
+-    spatial_radius/range_radius    = bandwidth parameters of truncated normal kernel. Controlling the size of the kernel determines the resolution of the mode detection.
+-    iter/eps                       = stopping criterion for meanshift procedure. The algorithm stops after iter iterations or if kernel center moves less than eps distance in an update step, whichever comes first.
+"""
+
+function meanshift{CT}(img::Array{CT, 2}, spatial_radius::Real, range_radius::Real; iter::Int = 50, eps::Real = 0.01)
+
+    rows, cols = size(img)
+    rowbins = Int(floor(rows/spatial_radius))
+    colbins = Int(floor(cols/spatial_radius))
+    colorbins = Int(floor(1/range_radius))
+
+    colorbin(x)::Int = min(floor(x/range_radius) + 1, colorbins)
+    rowbin(x)::Int = min(floor(x/spatial_radius) +1, rowbins)
+    colbin(x)::Int = min(floor(x/spatial_radius) +1, colbins)
+
+    buckets = Array{Vector{CartesianIndex{2}}}(rowbins, colbins, colorbins)
+    for i in CartesianRange(size(buckets))
+        buckets[i]=Array{CartesianIndex{2}}(0)
+    end
+
+    for i in CartesianRange(size(img))
+        push!( buckets[rowbin(i[1]), colbin(i[2]), colorbin(img[i])], i)
+    end
+
+    function dist2(a::SVector, b::SVector)::Float64
+        return ((a[1]-b[1])/spatial_radius)^2 + ((a[2]-b[2])/spatial_radius)^2 + ((a[3]-b[3])/range_radius)
+    end
+
+    function neighborhood_mean(pt::SVector{3, Float64})::SVector{3, Float64}
+        den = 0.0
+        num = SVector(0.0, 0.0, 0.0)
+
+        R = CartesianRange(size(buckets))
+        I1, Iend = first(R), last(R)
+        I = CartesianIndex(rowbin(pt[1]), colbin(pt[2]), colorbin(pt[3]))
+
+        for J in CartesianRange(max(I1, I-I1), min(Iend, I+I1))
+            for point in buckets[J]
```

Comment:
Out of curiosity, what's the advantage of using the buckets here? AFAICT you end up looping over every pixel anyway...

Presumbly you can check to see whether the spatially-closest pixel within a bucket is possibly within the required radius? And if not, exclude the entire bucket?

But even that is not enough justification for using the buckets. You could just say
```julia
for J in CartesianRange(Ipt-spatial_radius, Ipt+spatial_radius)
    ...
end
```
where `Ipt = CartesianIndex(round(Int, pt[1]), round(Int, pt[2]))` and only check points that might be within range.